File: Python/DRPerm.py
'''
Permutation Test for Distribution Shift via PO-risk(Pseudo-Outcome Risk) followed by the permute-then-refit procedure:

Hyperparameters:
- X:               Covariates array - (n_exist + n_new, p)
- Y:               Responses - (n_exist + n_new, )
- W:               Treatment(Batch) Assignment - (n_exist + n_new, )
- seed:            Random State Seeds
- n_split:         Number of splits
- clip_e:          Clipping value for the propensity score
- n_perm:          Number of permutations for the test
- alpha:           Significance Level for the test
- model_m:         Outcome Model
- model_e:         Propensity Score Model
Returns the cross-fitted value (m_hat, e_hat) evaluated on the whole data
as well as the fitted outcome model and propensity score models.
'''
#Specify the model registry factory so that we can fetch the model with flexibility afterwards:
model_registry = ModelRegistry(
  ntree = 150,
  ridge_alpha = 0.25,
  nthread = 1, maxit = 200, max_depth = 5,
  gamma = 0.25, eta = 0.15, mlp_hidden_size = 4,
  mlp_decay = 1e-5, mlp_max_iter = 500, mlp_trace = False,
  mlp_max_coef_reg = 10000, mlp_max_coef_clf = 10000,
  warn_xgb_labels = True, positive_class = 1
)
def DRPerm(
    X: np.ndarray, Y: np.ndarray, W: np.ndarray, *,
    seed: int = 0, n_splits: int = 5,
    clip_e = 0.01, n_perm = 150, alpha = 0.05, return_detail = True,
    model_m = 'rf_regression', model_e = 'rf_classification'):
    X = np.asarray(X)
    Y = _as_1d(Y)
    W = _as_1d(W).astype(int)
    n = X.shape[0]
    rng = np.random.default_rng(seed)
    # ----------------------------------------------------
    # Cross-fitted nuisance estimation:
    # ----------------------------------------------------
    folds = make_folds(n, n_folds = n_folds, seed = seed)
    mu_hat = np.zeros(n, dtype = float)
    e_hat  = np.zeros(n, dtype = float)
    for k, test_idx in enumerate(folds):
        train_idx = np.setdiff1d(np.arange(n), test_idx)
        X_train, X_test = X[train_idx], X[test_idx]
        Y_train = Y[train_idx]
        T_train = t[train_idx]
        #specify the outcome model and return it for each of the fold:
        model_propensity_score = model_registry[model_e]
        model_outcome = model_registry[model_m]
        fit_mu = model_outcome['fit'](
          X_train, Y_train, seed = seed + k
          )
        mu_hat[test_idx] = model_outcome['predict'](
          fit_mu, X_test
          )
        fit_e = model_propensity_score['fit'](
          X_train, T_train, seed = seed + 100 + k
          )
        e_hat[test_idx] = model_propensity_score['predict'](
          fit_e, X_test
          )
    e_hat = np.clip(e_hat, clip_e, 1 - clip_e)
    # ----------------------------------------------------
    # Calculate the Pseudo-Outcome Scores/PO-risk
    # ----------------------------------------------------
    residual_y = Y - mu_hat
    residual_t = T - e_hat
    pseudo_outcome = residual_y * residual_t
    # ----------------------------------------------------
    # Estimate the unpermuted PO-risk via outcome model
    # ----------------------------------------------------
    fit_tau = model_outcome['fit'](X, pseudo_outcome, seed = seed + 200 + k)
    tau_score = model_outcome['predict'](fit_tau, X)
    observed_po_risk = np.mean(tau_score ** 2)
    # ----------------------------------------------------
    # Permuted version of the PO-risk
    # ----------------------------------------------------
    permuted_po_risk = np.zeros(n_perm, dtype = float)
    for b in range(n_perm):
        t_perm = rng.permutation(t)
        fit_e_perm = model_propensity_score['fit'](
          X, t_perm, seed = seed + 300 + b
        )
        e_perm = model_propensity_score['predict'](
          fit_e_perm, X
        )
        e_perm = np.clip(e_perm, clip_e, 1 - clip_e)
        residual_t_perm = t_perm - e_perm
        pseudo_perm = residual_y * residual_t_perm
        fit_tau_perm = model_outcome['fit'](
          X, pseudo_perm, seed = seed + 400 + b
        )
        tau_perm = model_outcome['predict'](
          fit_tau_perm, X
        )
        permuted_po_risk[b] = np.mean(tau_perm ** 2)
    p_value = (1.0 + np.sum(permuted_po_risk >= observed_po_risk))/(n_perm + 1)
    out = {
      'statistic': float(observed_po_risk),
      'p_value': float(p_value),
      'reject': bool(p_value < alpha),
      'alpha': float(alpha),
      'model_outcome': model_outcome['name'],
      'model_propensity_score': model_propensity_score['name']
    }
    if return_detail:
        out.update({
          'mu_hat': mu_hat, 'e_hat': e_hat,
          'residual_y': residual_y, 'residual_t': residual_t,
          'pseudo_outcome': pseudo_outcome, 'tau_score': tau_score,
          'permuted_po_risk': permuted_po_risk
        })
    return out

File: R/DRPerm.R
#DRPerm R:
library(MASS)
library(glmnet)
library(ANN2)
library(CVST)
library(kernlab)
library(grf)
library(RandomProjectionTest)
library(caret)
source('model_registry.R')
source('utils.R')

####################################################################################
#'Permutation Test for Distribution Shift 
#'via Doubly Robust Pseudo-Outcome Learner
#
#'Implements a permute-then-refit permutation test based on the 
#'doubly robust pseudo-outcome risk. The procedure starts by
#'fitting the outcome model(on both treatment group and control group)
#'and propensity score model, then regress the pseudo-outcome
#'on the covariates. 
#
#'@param X             :Covariate Matrix, (n_exist + n_new, p)
#'@param Y             :Response Vectors, (n_exist + n_new, )
#'@param W             :Treatment(batch) Assignment Vectors, (n_exist + n_new, )
#'@param seed          :Random Seed, integer
#'@param n_splits      :Number of Splits.
#'@param clip_e        :Clipping Value for the propensity score
#'@param n_perm        :Number of permutations
#'@param alpha         :Significance Level alpha
#'@param m_model       :Outcome Model
#'@param e_model       :Propensity Score Model
#'@param tau_model     :Pseudo-Outcome Model
#'
#'@return A list containing the observed po-risk, permutation p-values
#'        as well as the rejection decisions.
####################################################################################
DRPerm <- function(X, Y, W,
    model_registry = model_registry, 
    m_model = 'rf_regression',
    e_model = 'logistic_classification',
    tau_model = 'rf_regression',
    seed = 2026, n_folds = 5,
    clip_e = 0.01, n_perm = 150, alpha = 0.05){
  n = nrow(X)
  p = ncol(X)
  nuisance_fit <- cross_nuisance_fit_po(X, Y, W,
    seed = seed, n_folds = n_folds,
    m_model = m_model, e_model = e_model)
  m0_hat = nuisance_fit$m0_hat
  m1_hat = nuisance_fit$m1_hat
  e_hat = nuisance_fit$e_hat
  pseudo_outcome <- (Y - ifelse(W == 1, m1_hat, m0_hat)) * (W - e_hat)/(e_hat * (1 - e_hat)) + (m1_hat - m0_hat)
  tau_po_model <- model_registry[[tau_model]]
  tau_fit <- fit_predict_oof(
    X = X, y = pseudo_outcome, folds = n_folds,
    model_spec = tau_po_model, seed = seed
    )$pred
  observed_po_risk <- mean((pseudo_outcome - tau_fit)**2)
  permuted_po_risk <- numeric(n_perm)
  for(b in 1:n_perm){
    W_perm = sample(W)
    nuisance_fit_b <- cross_nuisance_fit_po(X, Y, W_perm,
      seed = (seed + b^2), n_folds = n_folds,
      m_model = m_model, e_model = e_model)
    m0_hat_b = nuisance_fit_b$m0_hat
    m1_hat_b = nuisance_fit_b$m1_hat
    e_hat_b = nuisance_fit_b$e_hat
    pseudo_outcome_b <- (Y - ifelse(W_perm == 1, m1_hat_b, m0_hat_b)) * (W_perm - e_hat_b)/(e_hat_b * (1 - e_hat_b)) + (m1_hat_b - m0_hat_b)
    tau_po_model_b <- model_registry[[tau_model]]
    tau_fit_b <- fit_predict_oof(X = X, y = pseudo_outcome_b,
      folds = n_folds, model_spec = tau_po_model, seed = (seed + b^3))$pred
    permuted_po_risk[b] <- mean((pseudo_outcome_b - tau_fit_b) ** 2)
    print(b)
  }
  p_value <- (1 + sum(permuted_po_risk > observed_po_risk))/(1 + n_perm)
  return(
    list(
      observed_po_risk = observed_po_risk,
      permuted_po_risk = permuted_po_risk,
      p_value = p_value,
      rejected = (p_value < alpha)
        )
    )
}



model_registry <- default_model_registry()
#No Distribution Shift:
df_exist <- LM_generation(n = 1000, beta_hat = c(1,1,1,1,1,1,1,1), cor = 0.3, eps = 3)$df_return
df_new <- LM_generation(n = 1000, beta_hat = c(1,1,1,1,1,1,1,1), cor = 0.3, eps = 3)$df_return
W <- c(rep(0, nrow(df_exist)), rep(1, nrow(df_new)))
X <- rbind(df_exist[,2:(ncol(df_exist)-1)], df_new[,2:(ncol(df_new)-1)])
Y <- c(df_exist$Y, df_new$Y)
DRPerm(X, Y, W,
 model_registry = model_registry,
 n_perm = 25)$p_value


#Concept Drift:
df_exist <- LM_generation(n = 1000, beta_hat = c(1,1,1,1,1,1,1,1), cor = 0.3, eps = 3)$df_return
df_new <- LM_generation(n = 1000, beta_hat = c(1,1,0,0,0,0,0,0), cor = 0.3, eps = 3)$df_return
W <- c(rep(0, nrow(df_exist)), rep(1, nrow(df_new)))
X <- rbind(df_exist[,2:(ncol(df_exist)-1)], df_new[,2:(ncol(df_new)-1)])
Y <- c(df_exist$Y, df_new$Y)
DRPerm(X, Y, W,
 model_registry = model_registry,
 n_perm = 100)$p_value


#Covariate Shift(Mean Shift):
df_exist <- LM_generation(n = 1000, beta_hat = c(1,1,1,1,1,1,1,1), cor = 0.3, eps = 3)$df_return
df_new <- LM_generation(n = 1000, beta_hat = c(1,1,1,1,1,1,1,1), cor = 0.3, eps = 3, mean_X = 10)$df_return
W <- c(rep(0, nrow(df_exist)), rep(1, nrow(df_new)))
X <- rbind(df_exist[,2:(ncol(df_exist)-1)], df_new[,2:(ncol(df_new)-1)])
Y <- c(df_exist$Y, df_new$Y)
DRPerm(X, Y, W, 
  model_registry = model_registry,
  n_perm = 25)$p_value


#Covariate Shift(Variance Shift):
df_exist <- LM_generation(n = 1000, beta_hat = c(1,1,1,1,1,1,1,1), cor = 0.3, eps = 3)$df_return
df_new <- LM_generation(n = 1000, beta_hat = c(1,1,1,1,1,1,1,1), cor = 0.3, eps = 3, var_X = 5)$df_return
W <- c(rep(0, nrow(df_exist)), rep(1, nrow(df_new)))
X <- rbind(df_exist[,2:(ncol(df_exist)-1)], df_new[,2:(ncol(df_new)-1)])
Y <- c(df_exist$Y, df_new$Y)
DRPerm(X, Y, W, n_perm = 25)$p_value











